### ISSUE TITLE
Comparing dictionaries of differing structures fails

### ISSUE THREAD

**Issue by NxSoftware** [derekgreer/expectedObjects#34]
First of all great job on this library, it's incredibly useful!

I seem to have ran into a use case where I'm comparing arbitrarily nested dictionaries that may have different structures, specifically using dictionaries generated by using [YamlDotNet](https://github.com/aaubry/YamlDotNet) to parse YAML files.

Consider the following simple examples.

## Example 1

```csharp
var dictionary1 = new Dictionary<object, object>
{
    { "1", "val" },
    {
        "2", new Dictionary<object, object>
        {
            { "sub-1", "val" }
        }
    }
};
// Notice the keys "1" and "2" have been flipped
var dictionary2 = new Dictionary<object, object>
{
    { "2", "val" },
    {
        "1", new Dictionary<object, object>
        {
            { "sub-1", "val" }
        }
    }
};

dictionary1.ToExpectedObject().ShouldMatch(dictionary2);
```

Rather than failing gracefully and outputting the differences, it throws an `ArgumentException` with the following stack trace:

```
System.ArgumentException : Object must be of type String.
   at System.String.CompareTo(Object value)
   at ExpectedObjects.Strategies.ComparableComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.ReportEquality(Object expected, Object actual, String member)
   at ExpectedObjects.Strategies.KeyValuePairComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.AreEqual(Object expected, Object actual, String member)
   at ExpectedObjects.Strategies.EnumerableComparisonStrategy.<>c__DisplayClass1_1.<AreEqual>b__1(KeyValuePair`2 actualEntry)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at ExpectedObjects.Strategies.EnumerableComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.ReportEquality(Object expected, Object actual, String member)
   at ExpectedObjects.EqualityComparer.AreEqual(Object expected, Object actual, IWriter writer, Boolean ignoreTypeInformation)
   at ExpectedObjects.ExpectedObject.Equals(Object actual, IWriter writer, Boolean ignoreTypeInformation)
   at ExpectedObjects.ShouldExtensions.ShouldMatch[T](ExpectedObject expected, T actual)
```

## Example 2:

```csharp
var dictionary1 = new Dictionary<object, object>
{
    { "1", "val" },
    {
        "2", new Dictionary<object, object>
        {
            { "sub-1", "val" },
        }
    },
};
var dictionary2 = new Dictionary<object, object>
{
    { "1", "val" },
    { "2", 1 },
};

dictionary1.ToExpectedObject().ShouldMatch(dictionary2);
```

Throws:

```
System.InvalidCastException : Unable to cast object of type 'System.Int32' to type 'System.Collections.IEnumerable'.
   at ExpectedObjects.Strategies.EnumerableComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.ReportEquality(Object expected, Object actual, String member)
   at ExpectedObjects.Strategies.KeyValuePairComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.AreEqual(Object expected, Object actual, String member)
   at ExpectedObjects.Strategies.EnumerableComparisonStrategy.<>c__DisplayClass1_1.<AreEqual>b__1(KeyValuePair`2 actualEntry)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at ExpectedObjects.Strategies.EnumerableComparisonStrategy.AreEqual(Object expected, Object actual, IComparisonContext comparisonContext)
   at ExpectedObjects.EqualityComparer.Compare(Object expected, Object actual, String member, IWriter writer)
   at ExpectedObjects.EqualityComparer.ReportEquality(Object expected, Object actual, String member)
   at ExpectedObjects.EqualityComparer.AreEqual(Object expected, Object actual, IWriter writer, Boolean ignoreTypeInformation)
   at ExpectedObjects.ExpectedObject.Equals(Object actual, IWriter writer, Boolean ignoreTypeInformation)
   at ExpectedObjects.ShouldExtensions.ShouldMatch[T](ExpectedObject expected, T actual)
```

## Alternative Approach

As an alternative I have tried comparing an anonymous object against a dictionary but that doesn't seem to be supported as it appears to try to look for the properties on the dictionary rather than keys.

```csharp
var expectedObject = new
{
    key1 = "val",
    key2 = 1,
}.ToExpectedObject();

var dictionary = new Dictionary<object, object>
{
    { "key1", "val" },
    { "key2", 1 },
};

expectedObject.ShouldMatch(dictionary);
```

Output:

```
ExpectedObjects.ComparisonException : The expected object did not match the actual object.

The following issues were found:

1) Dictionary`2.key1:

  Expected:
    "val"
    
  Actual:
    member was missing
    


2) Dictionary`2.key2:

  Expected:
    1
    
  Actual:
    member was missing
```

**Comment by derekgreer** [comment]
ExpectedObjects uses a Chain-of-responsibility pattern for determining the comparison strategy to use for each member.  For dictionaries, one of either an ```EnumableComparisonStrategy``` (default) or ```OrdinalEnumerableComparisonStrategy``` is used.  

By default, ExpectedObjects compares every element of the expected enumerable with every element of the actual enumerable.  Of course, this was designed with the intention that the enumerable would yield the same data types and this is the issue you've uncovered.  I'll need to address this case ```EnumerableComparisonStrategy```, but for now you should be able to resolve your issue by using the ```OrdinalEnumerableComparisonStrategy``` like so:


```
            var expected = new Dictionary<object, object>
            {
                {"1", "val"},
                {
                    "2", new Dictionary<object, object>
                    {
                        {"sub-1", "val"}
                    }
                }
            };

            var actual = new Dictionary<object, object>
            {
                {"2", "val"},
                {
                    "1", new Dictionary<object, object>
                    {
                        {"sub-1", "val"}
                    }
                }
            };
        };

        _expected.ToExpectedObject(x => x.UseOrdinalComparison()).ShouldMatch(_actual);
```

**Comment by derekgreer** [comment]
Corrected in version 3.0.
